sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CallMethodViaReflection.scala

```diff
@@ -0,0 +1,166 @@
+
+  /** Name of the class -- this has to be called after we verify children has at least two exprs. */
+  @transient private lazy val className = children(0).eval().asInstanceOf[UTF8String].toString
+
+  /** True if the class exists and can be loaded. */
+  @transient private lazy val classExists = CallMethodViaReflection.classExists(className)
+
+  /** The reflection method. */
+  @transient lazy val method: Method = {
+    val methodName = children(1).eval(null).asInstanceOf[UTF8String].toString
+    CallMethodViaReflection.findMethod(className, methodName, argExprs.map(_.dataType)).orNull
+  }
+
+  /** A temporary buffer used to hold intermediate results returned by children. */
+  @transient private lazy val buffer = new Array[Object](argExprs.length)
+}
+
+object CallMethodViaReflection {
+  /** Mapping from Spark's type to acceptable JVM types. */
+  val typeMapping = Map[DataType, Seq[Class[_]]](
+    BooleanType -> Seq(classOf[java.lang.Boolean], classOf[Boolean]),
+    ByteType -> Seq(classOf[java.lang.Byte], classOf[Byte]),
+    ShortType -> Seq(classOf[java.lang.Short], classOf[Short]),
+    IntegerType -> Seq(classOf[java.lang.Integer], classOf[Int]),
+    LongType -> Seq(classOf[java.lang.Long], classOf[Long]),
+    FloatType -> Seq(classOf[java.lang.Float], classOf[Float]),
+    DoubleType -> Seq(classOf[java.lang.Double], classOf[Double]),
+    StringType -> Seq(classOf[String])
+  )
+
+  /**
+   * Returns true if the class can be found and loaded.
+   */
+  private def classExists(className: String): Boolean = {
+    try {
+      Utils.classForName(className)
+      true
+    } catch {
+      case e: ClassNotFoundException => false
+    }
+  }
+
+  /**
+   * Finds a Java static method using reflection that matches the given argument types,
+   * and whose return type is string.
+   *
+   * The types sequence must be the valid types defined in [[typeMapping]].
+   *
+   * This is made public for unit testing.
+   */
+  def findMethod(className: String, methodName: String, argTypes: Seq[DataType]): Option[Method] = {
+    val clazz: Class[_] = Utils.classForName(className)
+    clazz.getMethods.find { method =>
+      val candidateTypes = method.getParameterTypes
+      if (method.getName != methodName) {
+        // Name must match
+        false
+      } else if (!Modifier.isStatic(method.getModifiers)) {
+        // Method must be static
+        false
+      } else if (candidateTypes.length != argTypes.length) {
+        // Argument length must match
+        false
+      } else {
+        // Argument type must match. That is, either the method's argument type matches one of the
+        // acceptable types defined in typeMapping, or it is a super type of the acceptable types.
+        candidateTypes.zip(argTypes).forall { case (candidateType, argType) =>
+          typeMapping(argType).exists { acceptableType =>
+            acceptableType == candidateType || candidateType.isAssignableFrom(acceptableType)
```

Comment:
`candidateType.isAssignableFrom(acceptableType)` should already covers `acceptableType == candidateType`